astropy/table/tests/test_table.py

```diff
@@ -2208,6 +2208,18 @@ def test_table_meta_copy():
     assert t2.meta[1] is not t.meta[1]  # Value is NOT the same list object
 
 
+def test_init_table_with_names_and_structured_dtype():
+    """Test fix for #10393"""
+    arr = np.ones(2, dtype=np.dtype([('a', 'i4'), ('b', 'f4')]))
+    t1 = Table(arr)
+    for dtype in (arr.dtype, t1.dtype):
+        t2 = Table(names=['x', 'y'], dtype=arr.dtype)
+        assert t2.colnames == ['x', 'y']
+        assert str(t2['x'].dtype) == 'int32'
+        assert str(t2['y'].dtype) == 'float32'
+        assert len(t2) == 0
+
```

Comment:
```suggestion

    dt2 = np.dtype([('a', 'i2'), ('b', 'f8')]))
    t3 = Table(data=arr, names=['x', 'y'], dtype=dt2)
    assert t3.colnames == ['x', 'y']
    assert str(t3['x'].dtype) == 'int16'
    assert str(t3['y'].dtype) == 'float64'
    assert len(t3) == 2


```
With the proposed modifications to `table.py` this should work as well.

Reply:
Just getting back to this. l must be missing something, but I can't see how this is functionally different from the case of `has_data = True`. I'll go ahead and move this test function to `test_init_table.py`, but want to understand if your little snippet needs to go in as well.

Reply:
> Just getting back to this. l must be missing something, but I can't see how this is functionally different from the case of `has_data = True`

I missed that this still was a loose end. The test differs in the different `dtype` - `dtype=arr.dtype` is redundant per the behaviour we have now agreed on, since it already defaults to the input `arr`'s dtype. But for `t3` it is changed to `('int16', 'float64')`. Functionally this is equivalent to what ` test_set_dtype` is already testing, but that would use the `dtype=('i2', 'f8')` syntax. So I think it might be worth testing again with an `np.dtype` object here (and verify that `names=[]` still takes precedence for the colnames, which is perhaps less obvious and more easily broken).